### PR TITLE
[Python] Supports creating or editing resources.

### LINKED ISSUE #10822: [Improvement][Python] Python-gate supports creating or editing resources.
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/dolphinscheduler/issues?q=is%3Aissue) and found no similar feature requirement.

### Description

Our company now deploy and manage scripts automatically through Python-gate, but resource files need to be maintained manually.We find it inconvenient.So, I think Python-gate should supports creating or editing resources.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by hiSandog**
Our company now deploy and manage scripts automatically through Python-gate, but resource files need to be maintained manually.We find it inconvenient.So, I think Python-gate should supports creating or editing resources.

close: #10822

### CODE REVIEW

**Comment by zhongjiajie** [comment]
@hiSandog can we hold this PR for a while until we merge #10792? I think maybe we write code and pass the `Result` type is not the correct way, So in #10792 I migrate all `create` or `update` user and tenant code in `userService` and `TenantService`. In pythonGateway.java only call one function `tenantService.createTenantIfNotExists` which will keep python gateway code more easy to read

**Comment by hiSandog** [comment]
> @hiSandog can we hold this PR for a while until we merge #10792? I think maybe we write code and pass the `Result` type is not the correct way, So in #10792 I migrate all `create` or `update` user and tenant code in `userService` and `TenantService`. In pythonGateway.java only call one function `tenantService.createTenantIfNotExists` which will keep python gateway code more easy to read

ok

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/resource_definition.py
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/resource_definition.py]: I wonder, can we combine `name` and `suffix` to reduce our user's job? and extract the suffix in our python code or python gateway code?

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/resource_definition.py
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/resource_definition.py]: why we need `user` here? and how can we use `resource`? Does we have to use it alone instead of use it in the `process_definition`?

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/resource_definition.py
Comment (zhongjiajie) [reply dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/resource_definition.py]: And I think maybe we can directly use a path like `/dir1/dir2/file.py` in our `resource` class object, and inside our code(python or java code) we  parser it to `dir=/dir1/dir2`, `name=file` and `suffix=py`. which more easier to use in users side

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/python/PythonGateway.java
Comment (zhongjiajie) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/python/PythonGateway.java:560]: can we migrate all those main logic codes into https://github.com/apache/dolphinscheduler/blob/a6154220dc9fd78a894f9fc47c78378cba1d03cd/dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java#L1581 just like https://github.com/apache/dolphinscheduler/blob/a6154220dc9fd78a894f9fc47c78378cba1d03cd/dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/TenantServiceImpl.java#L352-L374

I find in we can add service in `ResourcesServiceImpl` will make our code easier to maintain. And with the pre review comment, maybe we only need to check whether the directory or file exists in https://github.com/apache/dolphinscheduler/blob/a6154220dc9fd78a894f9fc47c78378cba1d03cd/dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java#L329-L332 and create directory or file if not exists

**Comment by zhongjiajie** [comment]
restart the failed tests

File: dolphinscheduler-python/pydolphinscheduler/tests/core/test_resource_definition.py
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/tests/core/test_resource_definition.py]: A little typo here
```suggestion
def test_resource():
```

File: dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/process_definition.py
Comment (zhongjiajie) [inline dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/process_definition.py:66]:
```suggestion
    :param resource_list: Resource files required by the current process definition. You can create and modify
```

Comment (zhongjiajie) [reply dolphinscheduler-python/pydolphinscheduler/src/pydolphinscheduler/core/process_definition.py:66]: Great! thanks for adding this docstring

File: dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/ResourcesServiceTest.java
Comment (zhongjiajie) [inline dolphinscheduler-api/src/test/java/org/apache/dolphinscheduler/api/service/ResourcesServiceTest.java:609]: good to see you add test for service 👍

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java
Comment (zhongjiajie) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java]: also have `EMPTY_STRING` constants for `""`

**Comment by zhongjiajie** [comment]
BTW, I have review comment in https://github.com/apache/dolphinscheduler/pull/10823#discussion_r917614946. I think maybe we should better add a new interface named `createOrUpdateResource` in `ResourcesService.java`, just like you add `onlineCreateOrUpdateResourceWithDir` to `ResourcesService.java`.

And in `PythonGateway.java` we just add code below code to keep it as simple as possible, WDYT?

```java
    public Integer createOrUpdateResource(
            String userName, String fullName, String description, String resourceContent) {
        return resourceService.createOrUpdateResource(userName, fullName, description, resourceContent);
    }
```

**Comment by zhongjiajie** [comment]
I restart the failed E2E test

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/python/PythonGateway.java
Comment (zhongjiajie) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/python/PythonGateway.java:563]: I looks perfect now, thanks 👍

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java
Comment (zhongjiajie) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java]: please use constant `FOLDER_SEPARATOR` instead of bare `"/"`
```suggestion
        if (!fullName.startsWith(Coonstants.FOLDER_SEPARATOR)) {
```

Comment (zhongjiajie) [reply dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java]: same as other place

File: dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java
Comment (zhongjiajie) [inline dolphinscheduler-api/src/main/java/org/apache/dolphinscheduler/api/service/impl/ResourcesServiceImpl.java]: it is better to add `.` to `Constants` too

**Review by zhongjiajie — APPROVED** [review]
LGTM, thanks

**Comment by zhongjiajie** [comment]
well done, we currently supported create resource in python api now 👍 🎉 thanks again @hiSandog

**Pull request merged**
